docs/proposal/20210311-single-controller-multitenancy.md

```diff
@@ -0,0 +1,321 @@
+
+```go
+type VMContext struct {
+    *ControllerContext
+    VSphereVM   *v1alpha3.VSphereVM
+    PatchHelper *patch.Helper
+    Logger      logr.Logger
+    Session     *session.Session
+}
+```
+
+#### Proposed Changes
+
+The proposed changes below allow users to specify the account to use for clusters during runtime. The credentials can be provided by two approaches:
+
+- Referencing a cluster scoped `VsphereClusterAccount` to be used for credentials
+- Referencing a `Secret` that contains the credentials
+
+Changed Resources
+
+- `VsphereCluster`
+
+New Resources
+
+- cluster-scoped `VsphereClusterAccount` represents the vCenter account to use for reconciliation. This type should also contain references to namespaces that are allowed to use the account.
+
+Changes to VsphereCluster
+A new field is added to the VsphereClusterSpec to reference the `VsphereClusterAccount`. We intend to use `corev1.TypedLocalObjectReference` in order to ensure the only objects that can be referenced are either in the same namespace or are scoped globally.
+
+```go
+// VsphereClusterAccount is the account to be used for vcenter actions
+type VsphereClusterAccount struct {
+    metav1.TypeMeta `json:`",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec VsphereClusterAccountSpec `json:"spec,omitempty"`
+    Status VsphereClusterAccountStatus `json:"status, omitempty"`
+}
+
+type VsphereClusterAccountSpec struct {
+    // CredentialsRef references a Secret with the credentials to use
+    CredentialsRef *corev1.ObjectReference `json:"credentialsRef,omitempty"`
+
+    // AllowedNamespaces is used to identify which namespaces are allowed to use this account.
+    // Namespaces can be selected either by using an array of namespaces or with a label selector.
+    // If this object is nil, no namespaces will be allowed
+    //
+    // +optional
+    AllowedNamespaces *AllowedNamespaces `json:"allowedNamespaces"`
+}
+
+type AllowedNamespaces struct {
+    // If both NamespaceList and Selector are nil/empty, no namespaces will be allowed to use the account (default behaviour)
+    // If both NamespaceList and Selector are defined, the union of the two would be used to determine if a namespace is allowed to use the account
+    //
+    // Allowed namespaces listed out by name
+    // +optional
+    // +nullable
+    NamespaceList []string `json:"list"`
+
+    // Selector is a standard Kubernetes LabelSelector. A label query over a set of resources.
+    // +optional
+    Selector metav1.LabelSelector `json:"selector"`
+}
+
+type VsphereClusterSpec struct {
+    ...
+    // +optional
+    IdentityRef *corev1.TypedLocalObjectReference `json:"IdentityRef,omitempty"`
+}
+```
+
+#### Controller Changes
+
+- If IdentityRef is specified, and it references a `VsphereClusterAccount`, the CRD is fetched and used to create a session
```

Comment:
@randomvariable @fabriziopandini  -  it might worthwhile to inject this reference if nothing is set **and** there are `VsphereClusterAccount` that are allowed for that namespace. This avoids having to communicate the kind and Name to folks that want to provision clusters and that don't care much about creds ?

Reply:
I find the idea interesting, because it makes it easier to decouple roles and responsibilities
What there are more than one Account eligible for a cluster?

Reply:
We have a mechanism in CAPA for that called AWSClusterControllerIdentity that respects the existing behaviour as well as EC2 instance profiles for cross-account role assumption which is a key requirement for AWS. See https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/docs/proposal/20200506-single-controller-multitenancy.md#upgrade-strategy

Azure doesn't have cross-account role assumption so this wasn't an important use case for them, so was not implemented there.